app/src/main/scala/org/alephium/explorer/persistence/queries/ContractQueries.scala

```diff
@@ -0,0 +1,88 @@
+// (at your option) any later version.
+//
+// The library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the library. If not, see <http://www.gnu.org/licenses/>.
+
+package org.alephium.explorer.persistence.queries
+
+import scala.concurrent.ExecutionContext
+
+import slick.jdbc.{PositionedParameters, SetParameter, SQLActionBuilder}
+import slick.jdbc.PostgresProfile.api._
+
+import org.alephium.explorer.api.model.Pagination
+import org.alephium.explorer.persistence._
+import org.alephium.explorer.persistence.model.{CreateSubContractEventEntity, EventEntity}
+import org.alephium.explorer.persistence.schema.CustomGetResult._
+import org.alephium.explorer.persistence.schema.CustomSetParameter._
+import org.alephium.explorer.util.SlickUtil._
+import org.alephium.protocol.model.Address
+
+object ContractQueries {
+  def insertSubContractCreation(events: Iterable[EventEntity]): DBActionW[Int] = {
+    insertCreateSubContractEventEntities(
+      events.flatMap(CreateSubContractEventEntity.fromEventEntity)
+    )
+  }
+
+  private def insertCreateSubContractEventEntities(
+      events: Iterable[CreateSubContractEventEntity]): DBActionW[Int] = {
+    QuerySplitter.splitUpdates(rows = events, columnsPerRow = 6) { (events, placeholder) =>
+      val query =
+        s"""
+           |INSERT INTO create_sub_contract_events ("block_hash", "tx_hash", "contract", "sub_contract","block_timestamp", "event_order_in_block")
+           |VALUES $placeholder
+           |ON CONFLICT
+           | ON CONSTRAINT create_sub_contract_events_pk
+           | DO NOTHING
+           |""".stripMargin
+
+      val parameters: SetParameter[Unit] =
+        (_: Unit, params: PositionedParameters) =>
+          events foreach { event =>
+            params >> event.blockHash
+            params >> event.txHash
+            params >> event.contract
+            params >> event.subContract
+            params >> event.timestamp
+            params >> event.eventOrder
+        }
+
+      SQLActionBuilder(
+        queryParts = query,
+        unitPConv  = parameters
+      ).asUpdate
+    }
+  }
+
+  def getParentAddressQuery(subContract: Address)(
+      implicit ec: ExecutionContext): DBActionR[Option[Address]] = {
+    sql"""
+      SELECT contract
+      FROM create_sub_contract_events
+      WHERE sub_contract = $subContract
+      LIMIT 1
+      """.asASE[Address](addressGetResult).headOrNone
+  }
+
+  def getSubContractsQuery(contract: Address, pagination: Pagination): DBActionSR[Address] = {
+    sql"""
+      SELECT sub_contract
+      FROM create_sub_contract_events
+      WHERE contract = $contract
+      ORDER BY block_timestamp DESC, event_order_in_block
+      #${pagination.query}
```

Comment:
> Spliced pagination here would affect the query's ability to be used as prepared statement.

Great catch! Is it possible to reproduce this in a unit test?

Reply:
good point, I miss your issue #426 , really nice one. I guess it could be tackle on it's own. 
The schema changed a bit with my latest commits, I'll add the various idx.

Reply:
I did tried to add various indexes for timestamp, timestamp and event_order. But no one is ever used when performing the queries. 
As always with PG it depends with the amount of data, I tried on testnet but we still don't have too much contracts.
I think we could start without those indexes and retry in few monthes when we have more data